### DISCUSSION TITLE
Proxmox Backup Server

### DISCUSSION THREAD

**Discussion by DSDV** [tteck/Proxmox#2692] [Ideas]
Hi tteck, 
hi community, 

i recently had a few usecases for PBS on my PVE Systems. 
It is not the best solution regarding it as a "backup" since: 
> Caution
> 
> Installing the backup server directly on the hypervisor is not recommended. It is safer to use a separate physical server to store backups. Should the hypervisor server fail, you can still access the backups.

but there are still many [advantages](https://forum.proxmox.com/threads/what-are-the-advantages-of-pbs-vs-pve-backups.143798/#post-646506) to having it running on your PVE.

therefor i would like to ask if we could add an install option to the PVE Post Install Script 
or make a separate one to get PBS installed on your PVE: 
https://pbs.proxmox.com/docs/installation.html#install-proxmox-backup-server-on-proxmox-ve

with kind regards

PS: the PBS Post Install Script does not work if PBS is installed on a PVE system (:

**Comment by EMcoder888** [comment]
I use a Debian LXC running on my single node PVE with PBS installed. It has been stable for months backing up my other VM/LXCs to an external HDD. Maybe a script for a PBS LXC is a possibility.

**Reply by DSDV** [reply]
any reason why you did not install it on the node directly? 
in the proxmox forums it is recommended to do it that way instead of building a seperate vm.

**Reply by EMcoder888** [reply]
I originally had it installed on the PVE, but additional learning on reddit suggested to leave your PVE install as untouched as possible. So I removed it and created the LXC. The LXC could be moved to a new node if I one day had a second node to use in the future.

**Reply by DSDV** [reply]
that is interesting i only found staff members in the forum which said whats written in the documentation and pointed to direct install. 
the reasons iirc are especialy performance and iops where every bit seams to count. 
LXCs are nice but i fear in this case might add an unnecessary layer.

for moving the backups to another PBS there are sync tasks in the gui. 

(for me i sync everything to a nas anyway ^ ^)

**Comment by tteck** [comment]
I've never needed the Proxmox Backup Server. I back up 10 LXCs every 4 hours with a retention of 24 backups, providing me with 4 days' worth of backups at 4-hour intervals to choose from if needed. All backups are stored on a SAMSUNG T7 500GB Portable SSD, which currently has under 250GB of data stored. The complete backup process takes approximately 2 minutes in total.

![Screenshot 2024-03-25 11 09 02 AM](https://github.com/tteck/Proxmox/assets/21342670/5ec5f790-595a-4bef-96e9-7e94f1c2629a)

**Reply by DSDV** [reply]
to quote from my first link: 
> There is live restore where you are able to start and use a VM that isn't restored yet.
> There is encryption with zero-trust in mind.
> There are backup integrity checks.
> There is build-it sync jobs to replicate your backups between different PBS (offsite backups etc).
> In contrast to VZDump there is ransomware protection possible as you can set up privileges so that PVE is only able to create or restore backups but it isn't allowed to overwrite or delete any backups. All retention then could be handled by the PBS itself.
+ deduplication 

so i can have it all on pretty small drives is are advantages i really like (:

afaik the installation is not extraordinary difficult since it is adding repos and installing via apt.

but i think it would be nice to show up as option since also the option to restore single files could help a few folks (:

**Reply by tteck** [reply]
### EDITED
Your first link also quotes
>But running PBS on your PVE node isn't great:
>- you will lose that ransomware protection
>- unlike when using VZDump, there won'T be any backup archives/images you could directly restore to your PVE from your NAS. Without a running PBS all those backups on the NAS will be useless.
>- PBS needs massive IOPS performance (thats the reason why you should use SSDs and not HDDs to store your backups) and when using SMB/NFS the network latency will hurt that IOPS performance

I have created a Proxmox Backup Server LXC for testing.

```
bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/pbs.sh)"
```
 **NOTE:**
Set a root password if using autologin. This will be the PBS password.
```
sudo passwd root
```
Go to `IP:8007`

You can then execute the Proxmox Backup Server Post Install script in the PBS Shell.

**Reply by DSDV** [reply]
yep IOPS are a thing. 
for my use case i consider running on optane or nand flash so i think iops will not be a problem before i sync to nas. 

Thank you really much for the effort !

**Reply by tteck** [reply]
Do you think that a "Proxmox Backup Server LXC" script would be beneficial to the community?

**Reply by DSDV** [reply]
i think it has some features which add benefits especially in the storage department and are nice if you mess up while learning  / experimenting. 

but there are definitely drawbacks like the ones mentioned.
as well as the advice to "not have backups on the same device" which cant be overestimated in my eyes. 

so from my point of view i would say yes it could be beneficial but there must be warnings in place so folks don't get tempted to not have 'real' backup solutions and are aware of the risks (:

**Reply by EMcoder888** [reply]
@tteck Your directions for a PBS LXC match the notes I made when I used your Debian LXC and PBS post install scripts to create my own PBS LXC. It has been running well for about 3 months. 

The only additional steps I needed was mapping id's for my bind mount to add the backup user and group for accessing the external drive I used that was mounted on the PVE. I made a PVE vzdump of the install right after setup and plan to make periodic versions every few months in case I need to recover.

**Reply by tteck** [reply]
I updated my [post](https://github.com/tteck/Proxmox/discussions/2692#discussioncomment-8905351) above.

**Comment by bub4** [comment]
I could imagine that they don't recommend it because live and backup data should never be stored on the same hardware... 

I installed PBS as a VM on a second physical PVE node. The reason I used PBS is filebased restore and incremental backups.

But the most important reason for me was the ability to "short-term" boot the most important LXCs//VMs of the primary node in case of a hardware failure. There are however some zfs issues while updating the VM which I couldn't figure out to solve so far. But they seem not to influence the backup data.

**Comment by diegosps** [comment]
I have a home lab server with pve and I used the recent pbs lxc script to create the pbs.

Inside PBS, I can see all my three disks, including the external one which is not partitioned /sdc. 
Whenever I try to create a data storage in the empty disk, I get the following message:
Directory:
```
2024-04-05T19:00:33-03:00: create datastore 'backups' on disk sdc
2024-04-05T19:00:33-03:00: TASK ERROR: command "sgdisk" "-n1" "-t1:8300" "/dev/sdc" failed - status code: 2 - Problem opening /dev/sdc for reading! Error is 2.
The specified file does not exist!
```

ZFS:
```
2024-04-05T19:24:12-03:00: create Single zpool 'backups' on devices 'sdc'
2024-04-05T19:24:12-03:00: # "zpool" "create" "-o" "ashift=12" "-m" "/mnt/datastore/backups" "backups" "sdc"
2024-04-05T19:24:22-03:00: WARN: command "zpool" "create" "-o" "ashift=12" "-m" "/mnt/datastore/backups" "backups" "sdc" failed - status code: 1 - /dev/zfs and /proc/self/mounts are required.
Try running 'udevadm trigger' and 'mount -t proc proc /proc' as root.

2024-04-05T19:24:22-03:00: TASK ERROR: Error during 'zpool create', see task log for more details
```
I did not tried these commands.

In my syslogs this block repeats every 10 seconds. Searching the internet this looks pretty common.
```
Apr 05 19:32:38 pbs zed[846]: Failed to initialize libzfs
Apr 05 19:32:38 pbs systemd[1]: zfs-zed.service: Main process exited, code=exited, status=1/FAILURE
Apr 05 19:32:38 pbs systemd[1]: zfs-zed.service: Failed with result 'exit-code'.
Apr 05 19:32:38 pbs systemd[1]: zfs-zed.service: Scheduled restart job, restart counter is at 229.
Apr 05 19:32:38 pbs systemd[1]: Stopped zfs-zed.service - ZFS Event Daemon (zed).
Apr 05 19:32:38 pbs systemd[1]: Started zfs-zed.service - ZFS Event Daemon (zed).
Apr 05 19:32:38 pbs zed[857]: ZFS Event Daemon 2.2.3-pve1 (PID 857) 
```

I think I'm missing something really simple, but I don't know what. Is the lxc script 100%? If so, does anyone know what may be my problem?

**Reply by DSDV** [reply]
to be honest i still use the bare metal install ^.^
but on my new system (which currently only has 2vms migrated) i plan to use only tteck scrips and tools incl the PBS LXC. 
if i find some time next week ill try to prioritize it and test with a usb disk or so and report back (:

**Reply by diegosps** [reply]
I initialized the disk in pve and create a mount point in the privileged pbs lxc. This way I managed to create a data storage pointing to /mnt/backup.

I'm not sure this should be the right way, but the backup is working.
